java/arcs/core/testutil/Fuzzing.kt

```diff
@@ -0,0 +1,344 @@
+ *
+ * We could write such an invariant like this:
+ *
+ * invariant_withdrawals_lessThan_initialBalance_willApply(
+ *   initalBalance: A<Int>,
+ *   withdrawal: A<Withdrawal>
+ * ) {
+ *   val theInitialBalance = initialBalance()
+ *   val theWithdrawal = withdrawal()
+ *   assertThat(theWithdrawal.amount).isLessThan(theInitialBalance)
+ *   val account = Account(theInitialBalance)
+ *   assertThat(account.apply(theWithdrawal)).isTrue()
+ * }
+ *
+ * This is fine for unit testing, but not for fuzz testing as the initial assertion will fail
+ * quite regularly for randomly generated, uncorrelated inputs.
+ *
+ * We could filter out inputs that don't match the invariant rather than asserting, but with this
+ * approach, as the likelihood of randomly generating valid inputs gets smaller, tests either get
+ * less useful or take longer to run.
+ *
+ * Instead, we can represent the second input as dependent on the first:
+ *
+ * invariant_withdrawals_lessThan_initialBalance_willApply(
+ *   initalBalance: A<Int>,
+ *   withdrawal: T<Int, Withdrawal>
+ * ) {
+ *   val theInitialBalance = initialBalance()
+ *   val theWithdrawal = withdrawal(theInitialBalance)
+ *   assertThat(theWithdrawal.amount).isLessThan(theInitialBalance)
+ *   val account = Account(theInitialBalance)
+ *   assertThat(account.apply(theWithdrawal)).isTrue()
+ * }
+ *
+ * This isn't perfect in that the nature of the dependency still needs to be understood by
+ * calling code, but at least it lets the dependency exist in a meaningful way, without forcing
+ * a coupling between a specific pair of generators.
+ *
+ * The Function class (see below) allows the provision of explicit values when a T is expected.
+ */
+// TODO(shanestephens): Experiment with using A<Pair<I, O>> as input instead, and a generic
+// constructor for A<Pair<I, o>> given A<I> and T<I, O>.
+abstract class T<I, O> {
+  abstract operator fun invoke(i: I): O
+  fun asA(i: A<I>): A<O> {
+    class TAsA : A<O> {
+      override operator fun invoke(): O = this@T.invoke(i())
+    }
+    return TAsA()
+  }
+}
+
+/**
+ * A source of randomness.
+ */
+interface Seed {
+  fun nextDouble(): Double
+  fun nextLessThan(max: Int): Int
+  fun nextInRange(min: Int, max: Int): Int
+  fun nextInt(): Int
+}
+
+/**
+ * A seeded source of randomness. Will always produce the same sequence of values if given the
+ * same seed.
+ */
+open class SeededRandom(val seed: Int) : Seed {
+  val random = Random(seed)
+  override fun nextDouble(): Double = random.nextDouble()
+  override fun nextLessThan(max: Int): Int = random.nextInt(0, max)
+  override fun nextInRange(min: Int, max: Int): Int = random.nextInt(min, max + 1)
+  override fun nextInt(): Int = random.nextInt()
+  fun printSeed() {
+    println("Seed was $seed")
+  }
+}
+
+/**
+ * A source of randomness that is seeded by the current date & time.
+ */
+class DateSeededRandom : SeededRandom(LocalDateTime.now().hashCode())
```

Comment:
is ti simpler to use System.currentTimeMillis() that gives you a number directly?